Source/WebKit/GPUProcess/media/RemoteAudioDestinationManager.cpp

```diff
@@ -142,12 +143,9 @@ class RemoteAudioDestination final
             status = noErr;
         }
 
-        unsigned requestedSamplesCount = m_extraRequestedFrames;
-        for (; requestedSamplesCount < numberOfFrames; requestedSamplesCount += WebCore::AudioUtilities::renderQuantumSize) {
-            // Ask the audio thread in the WebContent process to render a quantum.
-            m_renderSemaphore.signal();
-        }
-        m_extraRequestedFrames = requestedSamplesCount - numberOfFrames;
+        ASSERT(m_frameCount->size() == sizeof(numberOfFrames));
+        memcpy(m_frameCount->data(), &numberOfFrames, sizeof(numberOfFrames));
```

Comment:
@kkinnunen-apple should probably comment but this doesn't look safe to me because it doesn't look like anything prevents the GPUProcess from writing into the frameCount shared memory while the WebProcess is in the middle of reading it. As such, the WebProcess could read corrupted data in some cases.

I believe we use `Atomics` to avoid such issues in other places in our code base. E.g.
```
struct Header {
        Atomic<ServerOffset> serverOffset;
        // Padding so that the variables mostly accessed by different processes do not share a cache line.
        // This is an attempt to avoid cache-line induced reduction of parallel access.
        // Use 128 bytes since that's the cache line size on ARM64, and enough to cover other platforms where 64 bytes is common.
        alignas(128) Atomic<ClientOffset> clientOffset;
    };

Header& header() const { return *reinterpret_cast<Header*>(m_sharedMemory->data()); }
```

That's from our streaming IPC implementation that Kimmo wrote so he is likely a good person to ask for help here.


Another more high level concern which may or may not be an issue in practice:
1. GPUProcess writes 128
2. GPUProcess writes 256
3. GPUProcess writes 128

but the WebProcess is a bit slow and took too long to process 1., as a result, by the time it processes the second quantum, 256 has already been overwritten by 128. So the WebProcess reads 128, 128, 128 and doesn't produce enough samples.

Reply:
Regarding corruption, I am relying on the fact that aligned memory won't be written from more than one thread and not at less than sizeof(register) at a time, so there should be no corrupted reads. But indeed the higher level race certainly exists.

Reply:
Atomic<> seems more appropriate in any case, thank you for alerting me of that, I was unaware of it

Reply:
We could add an IPC message from RemoteAudioDestination to RemoteAudioDestinationProxy containing the number of frames to read to address the higher level race but I'm not sure if that would add latency to the audio processing.

Reply:
Not sure if that would work but if you use `Atomic`, you might be able to:
1. Have the GPUProcess increment frameCount for every render call
2. Have the WebProcess exchange the frameCount with 0 atomically and use that frameCount to decide how much to render.

Reply:
That's a great idea. I'll try that.

Reply:
@cdumez I effectively took this suggestion, but instead of exchanging with 0 in the web process (which would involve writing to the shared memory from both processes), I kept a counter and handle overflow which should work out to be equivalent. This way only the GPU Process writes to the shared memory as I was not sure using atomic exchange on a shared memory location is safe.

I think this addresses the race, please let me know what you think.

Reply:
Yes, this way of dealing with the race seems OK to me.

It'd be good to switch to Atomic though.

Reply:
Great, I updated the PR to directly use `WTF::atomicLoad` and `WTF::atomicExchangeAdd` which calls into `WTF::Atomic` and also added a static assert that the atomic operations on uint32_t are lock free otherwise as far as I know they are not guaranteed to be atomic for shared memory